### ISSUE TITLE
Unsloth gradient checkpointing

### ISSUE THREAD

**Issue by kabachuha** [bghira/SimpleTuner#2504]
Unsloth gradient checkpointing is a custom algorithm by unsloth.

In Unsloth's blog it is claimed unsloth gradient checkpointing reduces VRAM usage by additional 30%, with a time overhead of ~2%.

https://unsloth.ai/blog/long-context

Aside from the Unsloth library (mainly for LLMs), this gradient checkpointing is implemented standalone in the other diffusion trainer - diffusion-pipe. I use unsloth checkpointing there by default and it's quite comfortable for me.

https://github.com/tdrussell/diffusion-pipe/blob/main/utils/unsloth_utils.py

Unsloth gradient checkpointing is licensed under LGPL, and it's compatible with this project's license.

If it works, would be great to have, especially for low VRAM systems and larger batch sizes.

**Comment by bghira** [comment]
avoiding copyright or any of the license differences by reimplementing it from their blog post in 2521 led to 2 it/sec training speed on a 5090, thanks for the request
